pkg/dockerregistry/server/digestcache.go

```diff
@@ -52,42 +75,134 @@ func (c digestToRepositoryCache) RepositoriesForDigest(dgst digest.Digest) []str
 	return repos.Copy()
 }
 
+func (c digestToRepositoryCache) RepositoryHasBlob(repo string, dgst digest.Digest) bool {
+	value, ok := c.Get(dgst.String())
+	if !ok {
+		return false
+	}
+	repos := value.(*repositoryBucket)
+	return repos.Has(repo)
+}
+
+// repositoryBucket contains a list of repositories with eviction timeouts.
 type repositoryBucket struct {
-	mu   sync.Mutex
-	list []string
+	mu    sync.Mutex
+	clock util.Clock
+	list  []bucketEntry
 }
 
 // Has returns true if the bucket contains this repository.
-func (i *repositoryBucket) Has(repo string) bool {
-	i.mu.Lock()
-	defer i.mu.Unlock()
-	for _, s := range i.list {
-		if s == repo {
-			return true
+func (b *repositoryBucket) Has(repo string) bool {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	b.evictStale()
+
+	return b.getIndexOf(repo) >= 0
+}
+
+// Add one or more repositories to this bucket.
+func (b *repositoryBucket) Add(ttl time.Duration, repos ...string) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	b.evictStale()
+	evictOn := b.clock.Now().Add(ttl)
+
+	for _, repo := range repos {
+		index := b.getIndexOf(repo)
+		arr := b.list
+
+		if index >= 0 {
+			// repository already exists, move it to the end with highest eviction time
+			entry := arr[index]
+			copy(arr[index:], arr[index+1:])
+			if entry.evictOn.Before(evictOn) {
+				entry.evictOn = evictOn
+			}
+			arr[len(arr)-1] = entry
+
+		} else {
+			// repo is a new entry
+			if len(arr) >= bucketSize {
+				copy(arr, arr[1:])
+				arr = arr[:bucketSize-1]
+			}
+			arr = append(arr, bucketEntry{
+				repository: repo,
+				evictOn:    evictOn,
+			})
 		}
+		b.list = arr
 	}
-	return false
 }
 
-// Add one or more repositories to this bucket.
-func (i *repositoryBucket) Add(repos ...string) {
-	i.mu.Lock()
-	defer i.mu.Unlock()
-	arr := i.list
+// Remove removes all the given repos from repository bucket.
+func (b *repositoryBucket) Remove(repos ...string) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
 	for _, repo := range repos {
-		if len(arr) >= bucketSize {
-			arr = arr[1:]
+		index := b.getIndexOf(repo)
+
+		if index >= 0 {
+			copy(b.list[index:], b.list[index+1:])
+			b.list = b.list[:len(b.list)-1]
+		}
+	}
+}
+
+// getIndexOf returns an index of given repository in bucket's array. If not found, -1 will be returned.
+func (b *repositoryBucket) getIndexOf(repo string) int {
+	arr := b.list
+
+	for i := 0; i < len(arr); i++ {
+		if arr[i].repository == repo {
+			return i
 		}
-		arr = append(arr, repo)
 	}
-	i.list = arr
+
+	return -1
 }
 
-// Copy returns a copy of the contents of this bucket in a threadsafe fasion.
-func (i *repositoryBucket) Copy() []string {
-	i.mu.Lock()
-	defer i.mu.Unlock()
-	out := make([]string, len(i.list))
-	copy(out, i.list)
+// evictStale removes stale entries from the list and shifts all the survivalists to the front.
+func (b *repositoryBucket) evictStale() {
+	arr := b.list
```

Comment:
saw this method of filtering a list at https://github.com/golang/go/wiki/SliceTricks#filtering-without-allocating, and I can't decide if I like it or not:
```
results := b.list[:0]
for _, entry := range b.list {
  if now.Before(entry.evictOn) {
    results = append(results, entry)
  }
}
b.list = results
```